### ISSUE TITLE
Issues on macOS on Apple SoC (M1 Max)

### ISSUE THREAD

**Issue by tingspain** [NedoProgrammer/TinyDialogsNet#4]
I downloaded the original [TinyDialogs](https://sourceforge.net/projects/tinyfiledialogs/) v3.9.0 and compiled it to have the ARM version of the library 

Then I updated all the runtimes on your projects and compiled it, and it worked flawless. However I found two bugs running it on my macbook ( OSX Ventura 13.2 - M1 Max)

These are the two bugs:  

**1.- Issue With Titles and Descriptions (any string):**
As you can see from the screenshot, only the first character of the strings are taken into account. 
 
**CODE:**
````c#
var response = Dialogs.MessageBox(Dialogs.MessageBoxButtons.YesNo,
                                  Dialogs.MessageBoxIconType.Error,
                                  Dialogs.MessageBoxDefaultButton.OkYes,
                                  "Unhandled exception!",
                                  "Would you like to submit a crash report?");
````

**SCREENSHOT**
<img width="446" alt="image" src="https://user-images.githubusercontent.com/1858083/222576134-3679e709-d891-415c-a10f-6e19673a3533.png">

**2.- Issue With  Array of Strings (ex: filterPatterns or filters):**
I had issues with any array of stings like filters at tinyfd_saveFileDialog, or filterPatterns at tinyfd_openFileDialog.  As it can be seen in the screenshot, the dialog is not applying properly the filter.  

**CODE:**
````c#
var files2 = Dialogs.OpenFileDialog("Please select an image..", "", new[] { "*.png" }, "PNG images", true);

````

**SCREENSHOT**
<img width="981" alt="image" src="https://user-images.githubusercontent.com/1858083/222578467-3e965ea1-bc57-491d-bdea-e5734b949d64.png">

**Solution**
I tried different solution but this was the most clean for me. I basically set the charset of each import to CharSet.Auto, this fixed the simple string and the array of strings issues leaving the import code as follows:

``` c#
private const string EntryName = "tinyfiledialogs";

[DllImport(EntryName, CallingConvention = CallingConvention.Cdecl)]
private static extern void tinyfd_beep();

[DllImport(EntryName, CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Auto)]
private static extern int tinyfd_notifyPopup( string title,  string message, string iconType);

[DllImport(EntryName, CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Unicode)]
private static extern int tinyfd_messageBox( string title,  string message, string dialogType, string iconType,
		int defaultButton);

[DllImport(EntryName, CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Auto)]
private static extern IntPtr tinyfd_inputBox( string title,  string message,  string defaultInput);

[DllImport(EntryName, CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Auto)]
private static extern IntPtr tinyfd_saveFileDialog( string title,  string defaultPath, int numOfFilterPatterns,
		string[] filters,  string filterDescription);

[DllImport(EntryName, CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Auto)]
private static extern IntPtr tinyfd_openFileDialog(string title, string defaultPath, int numOfFilterPatterns,
		string[] filterPatterns, string filterDescription, int allowMultipleSelects);

[DllImport(EntryName, CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Auto)]
private static extern IntPtr tinyfd_colorChooser( string title,  string defaultHex, byte[] defaultRgb,
		byte[] oResultRgb);

[DllImport(EntryName, CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Auto)]
private static extern IntPtr tinyfd_selectFolderDialog( string title,  string defaultPath);

```

I didnt try yet this solution on Linux or Windows, so I dont know if it will work on those two platforms too. If not, then we should need to use some preprocessor directives checks as follows:

``` c#
#if OS_WINDOWS
  // Windows-specific code
 //...
[DllImport(EntryName, CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Unicode)]
private static extern IntPtr tinyfd_saveFileDialog( string title,  string defaultPath, int numOfFilterPatterns,
		string[] filters,  string filterDescription);

[DllImport(EntryName, CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Unicode)]
private static extern IntPtr tinyfd_openFileDialog(string title, string defaultPath, int numOfFilterPatterns,
		string[] filterPatterns, string filterDescription, int allowMultipleSelects);
//...
#elif OS_MAC
  // Mac-specific code
 //...
[DllImport(EntryName, CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Ansi)]
private static extern IntPtr tinyfd_saveFileDialog( string title,  string defaultPath, int numOfFilterPatterns,
		string[] filters,  string filterDescription);

[DllImport(EntryName, CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Ansi)]
private static extern IntPtr tinyfd_openFileDialog(string title, string defaultPath, int numOfFilterPatterns,
		string[] filterPatterns, string filterDescription, int allowMultipleSelects);
//...
#elif OS_LINUX
  // Linux-specific code
#elif OS_FREEBSD
  // FreeBSD-specific code
#endif
```

Where the OS-specific constants can be defined at the project file (*.csproj) as follows:

````
<PropertyGroup Condition="$([MSBuild]::IsOSPlatform('Windows'))">
  <DefineConstants>OS_WINDOWS</DefineConstants>
</PropertyGroup>
<PropertyGroup Condition="$([MSBuild]::IsOSPlatform('Linux'))">
  <DefineConstants>OS_LINUX</DefineConstants>
</PropertyGroup>
<PropertyGroup Condition="$([MSBuild]::IsOSPlatform('FreeBSD'))">
  <DefineConstants>OS_FREEBSD</DefineConstants>
</PropertyGroup>
<PropertyGroup Condition="$([MSBuild]::IsOSPlatform('OSX'))">
  <DefineConstants>OS_MAC</DefineConstants>
</PropertyGroup>

````

**Comment by tingspain** [comment]
I have another Intel Macbook. I will try this solution on that machine and i will report it if it works too.

**Comment by vareille** [comment]
I don't know if it helps, but in tinyfiledialogs v3.13.3 I have added the AppleSilicon dylib file (previously it was only Intel).
I've also added Linux .so files for x86 and x64.
In tinyfiledialogs, I have a C# file called tinyfiledialogsTest.cs  which separates UTF-8 and UTF-16 calls. I have the impression that my C# file wasn't taken into account in this project.

**Comment by nedoxff** [comment]
hello! forgive me for responding after such a long time.

i've been thinking about rewriting this project from the ground up for some time now. the goal of the library was to abstract the interop code and UTF conversions away from the end-user. i didn't do it in the best way possible and it was *good enough* for my personal needs at the time, and i didn't expect a lot of people to actually use it, sorry. 😅

i hope i'll be able to refactor the code after i get some things sorted out. i'll update the native libraries, however unfortunately i don't own any apple devices to actually test the code, but i'll try to find a workaround.

**Comment by nedoxff** [comment]
i managed to replicate the bug on an Intel OSX virtual machine. hopefully this should be fixed with the 2.0.0 release!
unfortunately i'm having some troubles with updating the nuget package, but they should be resolved soon.
if the bug persists, feel free to reopen the issue!

**Comment by nedoxff** [comment]
[the 2.0.0 version is now published on nuget.](https://www.nuget.org/packages/TinyDialogsNet/2.0.0)
